src/libs/ajax/DatasetBuilder.ts

```diff
@@ -1,11 +1,92 @@
+// Types that can be used to create a criteria.
+
+export interface DomainType {
+  id: number;
+  category: string;
+  values: string[];
+}
+
+type DataType = 'range' | 'list';
+
+export interface ProgramDataType {
+  id: number;
+  name: string;
+  dataType: DataType;
+}
+
+export interface ProgramDataRangeType extends ProgramDataType {
+  min: number;
+  max: number;
+}
+
+export interface ProgramDataListTypeValue {
+  id: number;
+  name: string;
+}
+
+export interface ProgramDataListType extends ProgramDataType {
+  values: ProgramDataListTypeValue[];
+}
+
+// discriminating union
+// define the three types
+export type CriteriaType = DomainType | ProgramDataRangeType | ProgramDataListType;
```

Comment:
I think this type is only used in the UI code rather than for Ajax requests, should we co-locate it? (I am happy either way - I know you moved a bunch of stuff here already)